### PR TITLE
Redesigned cpp client.

### PR DESCRIPTION
**Pull request by paweljaciow**

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+# *************** OpenSSL **************** #
+set(OPENSSL_USE_STATIC_LIBS TRUE)
+find_package(OpenSSL REQUIRED)
+add_library(ssl_pkg INTERFACE IMPORTED)
+set_property(TARGET ssl_pkg
+        PROPERTY INTERFACE_LINK_LIBRARIES ${OPENSSL_LIBRARIES})
+
+# *************** GRPC **************** #
+find_package(Grpc REQUIRED)
```
Comment (tomjad) [inline cpp/CMakeLists.txt]: Set `Grpc_ROOT` before `find_package(Grpc...)`, it'll be easier to change default grpc path here than looking for it inside cmake module (although you should left default value inside module as it is now)

File: cpp/libsarmata-client/sarmata_client.cpp [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    auto stub = ASR::NewStub(grpc::CreateChannel(service_address_, grpc::InsecureChannelCredentials()));
+
+    auto stream = stub->Recognize(&context);
+
+    const auto requests = build_request(config, audio_sample_rate_hz, audio_byte_content);
+
+    const auto& config_request = requests.front();
+    stream->Write(config_request);
+
+    std::atomic<bool> half_closed_stream{false};
```
Comment (tomjad) [inline cpp/libsarmata-client/sarmata_client.cpp]: please add some comments what for `half_closed_stream`  is required

File: cpp/libsarmata-client/sarmata_client.cpp [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+    if (is_error) {
+        const auto& error = response.error();
+        std::cerr << "Received error response: (" << ResponseStatus_Name(status) << ") " << error << std::endl;
+    }
+
+    return is_error;
+}
+
+std::string grpc_status_to_string(const grpc::Status& status) {
```
Comment (tomjad) [inline cpp/libsarmata-client/sarmata_client.cpp:205]: I think that extracting GRPC-related functions (`build_recognition_config`, `build_request`, `error_response` and `grpc_status_to_string`) to separate source file would make `sarmata_client.cpp` easier to read

File: cpp/libsarmata-client/sarmata_client.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
+#ifndef __SARMATA_CLIENT_H__
+#define __SARMATA_CLIENT_H__
+
+#include "sarmata_asr.grpc.pb.h"
+
+
+namespace techmo { namespace sarmata {
+
+struct SarmataClientConfig {
```
Comment (tomjad) [inline cpp/libsarmata-client/sarmata_client.h]: Could we change the name of this struct to something more related to ASR? This is not a config of SarmataClient, maybe naming it `RequestConfig` ? or `ASRSessionConfig`? would be more appropriate?

File: cpp/libsarmata-client/sarmata_client.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    std::string session_id = "";        // Session ID to be passed to the service. If not specified, the service will generate a default session ID itself.
```
Comment (tomjad) [inline cpp/libsarmata-client/sarmata_client.h]: Enhance comment. Session ID is the best way to match log's from client application with these on server side

File: cpp/libsarmata-client/sarmata_client.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    std::string service_settings = "";  // Semicolon-separated list of key=value pairs defining settings to be sent to service via gRPC request.
```
Comment (tomjad) [inline cpp/libsarmata-client/sarmata_client.h]: Add info that full list of accepted settings is placed on github/ doc directory

File: cpp/sarmata-client/main.cpp [hunk 7]
```diff
@@ -137,33 +87,38 @@ po::options_description CreateOptionsDescription(void) {
...
             return 0;
         }
+        po::notify(userOptions);
+
+        const techmo::sarmata::SarmataClientConfig config = CreateSarmataClientConfig(userOptions);
+        techmo::sarmata::SarmataClient sarmata_client{ userOptions["service-address"].as<std::string>() };
 
         if (userOptions.count("define-grammar")) {
-            return defineGrammar(userOptions);
+            if (config.grammar_name.empty()) {
```
Comment (tomjad) [inline cpp/sarmata-client/main.cpp]: do some additional logging - which mode we are running here (defining grammar, deleting grammar)
consider extracting this as a separate method (or even add factory to create selected mode - define-grammar, run-recognition)

**Review by tomjad — CHANGES_REQUESTED** [review]
Looks much better than previous version. Few minor improvements required and then we can 'release' it

File: cpp/libsarmata-client/sarmata_client.cpp (hunk 3 above)
Comment (paweljaciow) [reply cpp/libsarmata-client/sarmata_client.cpp:205]: I don't agree that a separate file would improve readability here. What I can offer instead is to move public methods (`DefineGrammar` and `Recognize`) to the top of the file.

File: cpp/libsarmata-client/sarmata_client.h (hunk 4 above)
Comment (paweljaciow) [reply cpp/libsarmata-client/sarmata_client.h]: `SarmataSessionConfig` would be appropriate. `ASR-` could be confused with the other ASR product. Also, it is used in the server's code. And `RequestConfig` is already in a request.

File: cpp/sarmata-client/main.cpp (hunk 7 above)
Comment (paweljaciow) [reply cpp/sarmata-client/main.cpp]: I think separate methods would be enough.

**Review by tomjad — APPROVED** [review]

**Pull request merged**